plugin/evm/block.go

```diff
@@ -102,14 +107,77 @@ func (b *Block) syntacticVerify() error {
+func (b *Block) ShouldVerifyWithContext(context.Context) (bool, error) {
+	precompileConfigs := b.vm.chainConfig.AvalancheRules(b.ethBlock.Number(), b.ethBlock.Timestamp()).Precompiles
+	for _, tx := range b.ethBlock.Transactions() {
+		for _, accessTuple := range tx.AccessList() {
+			if _, ok := precompileConfigs[accessTuple.Address]; ok {
+				log.Debug("Should verify block with proposer VM context", "block", b.ID(), "height", b.Height())
+				return true, nil
+			}
+		}
+	}
+	log.Debug("Block does not require proposer VM context for verification.", "block", b.ID(), "height", b.Height())
+
+	return false, nil
+}
+
+// VerifyWithContext implements the block.WithVerifyContext interface
+func (b *Block) VerifyWithContext(ctx context.Context, proposerVMBlockCtx *block.Context) error {
+	if proposerVMBlockCtx != nil {
+		log.Debug("Verifying block with context", "block", b.ID(), "height", b.Height())
+	} else {
+		log.Debug("Verifying block without context", "block", b.ID(), "height", b.Height())
+	}
+
+	// The engine may call VerifyWithContext multiple times on the same block with different contexts.
+	// Since the engine will only call Accept/Reject once, we need to ensure that we only call verify with writes=true once.
+	// Additionally, if a block is already in processing, then it has already passed verification and we only need to check that the predicates are
+	// still valid in the different context to ensure that Verify should return nil.
+	if b.vm.State.IsProcessing(b.id) {
+		rules := b.vm.chainConfig.AvalancheRules(b.ethBlock.Number(), b.ethBlock.Timestamp())
+		predicateCtx := precompile.PredicateContext{
+			SnowCtx:            b.vm.ctx,
+			ProposerVMBlockCtx: proposerVMBlockCtx,
+		}
+		for _, tx := range b.ethBlock.Transactions() {
+			if err := core.CheckPredicates(rules, &predicateCtx, tx); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	return b.verify(proposerVMBlockCtx, true)
 }
 
-func (b *Block) verify(writes bool) error {
+// Verify the block, including checking precompile predicates
+func (b *Block) verify(proposerVMBlockCtx *block.Context, writes bool) error {
 	if err := b.syntacticVerify(); err != nil {
 		return fmt.Errorf("syntactic block verification failed: %w", err)
 	}
 
+	// Only enforce predicates if the chain has already bootstrapped.
+	// If the chain is still bootstrapping, we can assume that all blocks we are verifying have
+	// been accepted by then network and the predicate was validated by the network when the
+	// block was verified.
+	if b.vm.bootstrapped {
+		rules := b.vm.chainConfig.AvalancheRules(b.ethBlock.Number(), b.ethBlock.Timestamp())
+		predicateCtx := precompile.PredicateContext{
+			SnowCtx:            b.vm.ctx,
+			ProposerVMBlockCtx: proposerVMBlockCtx,
+		}
+
+		for _, tx := range b.ethBlock.Transactions() {
+			if err := core.CheckPredicates(rules, &predicateCtx, tx); err != nil {
+				return err
+			}
+		}
```

Comment:
Looks like this exact code block is duplicated in `verify` and `VerifyWithContext`. May want to consider moving it to a helper to avoid the duplication.